embedchain/vectordb/lancedb.py

```diff
@@ -0,0 +1,281 @@
+from typing import Any, Dict, List, Optional, Union
+
+import pyarrow as pa
+
+try:
+    import lancedb
+except ImportError:
+    raise ImportError("LanceDB is required. Install with `pip install embedchain[lancedb]`") from None
+
+from embedchain.config.vectordb.lancedb import LanceDBConfig
+from embedchain.helpers.json_serializable import register_deserializable
+from embedchain.vectordb.base import BaseVectorDB
+
+
+@register_deserializable
+class LanceDB(BaseVectorDB):
+    """
+    LanceDB as vector database
+    """
+
+    BATCH_SIZE = 100
+
+    def __init__(
+        self,
+        config: Optional[LanceDBConfig] = None,
+    ):
+        """LanceDB as vector database.
+
+        :param config: LanceDB database config, defaults to None
+        :type config: LanceDBConfig, optional
+        """
+        if config:
+            self.config = config
+        else:
+            self.config = LanceDBConfig()
+
+        self.client = lancedb.connect(self.config.dir or "~/.lancedb")
+
+        super().__init__(config=self.config)
+
+    def _initialize(self):
+        """
+        This method is needed because `embedder` attribute needs to be set externally before it can be initialized.
+        """
+        self._get_or_create_collection(self.config.collection_name)
+
+    def _get_or_create_db(self):
+        """
+        Called during initialization
+        """
+        return self.client
+
+    def _generate_where_clause(self, where: Dict[str, any]) -> str:
+        """
+        This method
+        """
+
+        where_filters = ""
+
+        if len(list(where.keys())) == 1:
+            where_filters = f"{list(where.keys())[0]} = {list(where.values())[0]}"
+            return where_filters
+
+        where_items = list(where.items())
+        where_count = len(where_items)
+
+        for i, (key, value) in enumerate(where_items, start=1):
+            condition = f"{key} = {value} AND "
+            where_filters += condition
+
+            if i == where_count:
+                condition = f"{key} = {value}"
+                where_filters += condition
+
+        return where_filters
+
+    def _get_or_create_collection(self, table_name: str, reset=False):
+        """
+        Get or create a named collection.
+
+        :param name: Name of the collection
+        :type name: str
+        :return: Created collection
+        :rtype: Collection
+        """
+        if not self.embedder:
+            raise ValueError("Embedder not set. Please set an embedder with `set_embedder` before initialization.")
+
+        schema = pa.schema(
+            [
+                pa.field("vector", pa.list_(pa.float32(), list_size=self.embedder.vector_dimension)),
+                pa.field("doc", pa.string()),
+                pa.field("metadata", pa.string()),
```

Comment:
Wouldn't metadata be a Dict here?

Reply:
We care converting dict to string

Reply:
Yes that's my question can't we keep in dict format, because when `query` function is called it's not getting converted back again to dict.

Reply:
No putting and extracting in dict is not possible